### DISCUSSION TITLE
What does pc in pc_bis_Element_CodeValue stand for？

### DISCUSSION THREAD

**Discussion by Snailya** [iTwin/itwinjs-core#3338] [Q&A] [answered]
so does rc in rc_generic_PhysicalObject_bis_PhysicalModel_bis_PhysicalPartition_ncc_bis_RepositoryLink？
Is there any article about the naming rule for these keys?
![image](https://user-images.githubusercontent.com/19658624/157818465-41d411dc-c0d2-42b3-a8e9-99516513de7b.png)

**Comment by ColinKerr** [comment, marked as answer]
'rc_' stands for related content
generic_PhysicalObject, bis_PhysicalModel, bis_PhysicalPartition and bis_RepositoryLink are classes in the format <schema>_<class> e.g. the PhysicalObject class from the schema Generic.  
'ncc_' stands for nested content

These keys are generated by our presentation library (https://www.itwinjs.org/presentation/).

If you explain what are you trying to do with this key we can provide some guidance.

**Reply by Snailya** [reply]
I'm trying to learn how to use ruleset to custom presentation. 
I'm struggled with the doc, it seems that learing curve is step and preliminary knowledge is needed. 
So I log content with rule details changes. 
Could you provide any link or road map for better understanding?

**Reply by Snailya** [reply]
![image](https://user-images.githubusercontent.com/19658624/158337100-7cdf4c57-f2e3-4780-b894-742190e58d95.png)
Could you give an sample code of presentation rule on displaying property defined in GeometricElement3d using a PropertyGrid.

**Reply by grigasp** [reply]
Most of these properties are marked hidden with a custom attribute in BisCore schema, e.g. [see this](https://github.com/iTwin/bis-schemas/blob/6b1716f1da40d40b9c29764a33317f6a568e7cc3/Domains/Core/Released/BisCore.01.00.14.ecschema.xml#L1098). This means that by default the presentation library ignores them. You can override that using a content modifier with [a property display override](https://github.com/iTwin/itwinjs-core/blob/master/docs/presentation/Content/ContentModifier.md#attribute-propertyoverrides).

Also note that we're not going to return `GeometryStream` as there's no component that can display its binary value in a meaningful way.

**Reply by Snailya** [reply]
Thanks, I displayed all these hidden property with a custom ruleset like:
`{
  id: 'default-ae-ruleset',
  rules: [{
    ruleType: RuleTypes.Content,
    specifications: [{
      specType: ContentSpecificationTypes.SelectedNodeInstances
    }]
  }, {
    ruleType: RuleTypes.ContentModifier,
    propertyOverrides: [{
      name: '*',
      isDisplayed: true
    }]
  }]
}`
Is there any built-in function that could calculate the volume of the selected element?

**Reply by grigasp** [reply]
Glad it worked.

Regarding the volume, it depends on whether you're okay with getting volume of the bounding box or need an accurate volume of the element itself. In case of the latter, I think it can get quite complex as it depends on the geometry. Adding @pmconne who should know more on this.

**Reply by Snailya** [reply]
![image](https://user-images.githubusercontent.com/19658624/158543643-240e3b52-2b2f-48ef-806a-a9e029a11b47.png)
Do you know which volume MicroStation use?

**Reply by pmconne** [reply]
MicroStation calculates properties like volume, area, and length on the fly from the element geometry. These properties are not stored.
You can obtain such properties in iTwin.js via [IModelConnection.getMassProperties](https://www.itwinjs.org/reference/core-frontend/imodelconnection/imodelconnection/#getmassproperties).
You cannot obtain them via ECSql (and by extension, ECPresentation rules) because ECSql does not look inside element geometry.

**Reply by Snailya** [reply]
I've configured that the primative properties could be removed using propertyoverrides. But what about related properties?
How to configure ruleset to remove the rc_generic_PhysicalObject_bis_PhysicalModel_bis_PhysicalPartition_ncc_bis_RepositoryLink of Displayvalues?
![image](https://user-images.githubusercontent.com/19658624/159452642-e426d88e-20f8-4d42-9e90-1564f2d93be7.png)

**Reply by grigasp** [reply]
The related properties come from the [BisCore supplemental ruleset](https://github.com/iTwin/itwinjs-core/blob/master/presentation/backend/assets/supplemental-presentation-rules/BisCore.PresentationRuleSet.json). Generally, you're supposed to be able to override related property specifications by adding similar rules with higher priority, which override the ones in supplemental ruleset. Sadly, that seems to not work when trying to completely disable related property specifications. I created a work item in our internal backlog and we'll work on a fix for this asap.

In the meantime, you could still exclude related properties by supplying a [fields selector](https://github.com/iTwin/itwinjs-core/blob/cc2abab4f035283a66a0e21a9e6590932769baea/presentation/common/src/presentation-common/content/Descriptor.ts#L187) with `type: "exclude"` when requesting content.

**Reply by Snailya** [reply]
Could you give a sample on how to use fields selector?
I tried with code below, but nothing happened.
![image](https://user-images.githubusercontent.com/19658624/159604448-e250bd70-c831-45fb-8c6a-c739af7ba2bc.png)

**Reply by grigasp** [reply]
Something like this should work:
```ts
class MyPropertyDataProvider extends PresentationPropertyDataProvider {
    protected override async getDescriptorOverrides(): Promise<DescriptorOverrides> {
        const baseOverrides = await super.getDescriptorOverrides();
        const descriptor = await this.getContentDescriptor();
        if (!descriptor)
            return baseOverrides;
        return {
            ...baseOverrides,
            fieldsSelector: {
                type: "exclude",
                fields: descriptor.fields
                    .filter((field) => field.isNestedContentField())
                    .map((field) => field.getFieldDescriptor()),
            }
        }
    }
}
```

**Comment by grigasp** [comment]
Just to add to what @ColinKerr already said... Please don't try to parse any of that information from field's name - that's an internal detail that might change without notice. Let us know what you're trying to achieve and we'll suggest the right approach.
